services/typescript-service.ts

```diff
@@ -0,0 +1,316 @@
+	public transpileWithDefaultOptions(projectDir: string, options?: ITypeScriptTranspileOptions): IFuture<string> {
+		return ((): string => {
+			options = options || {};
+			let compilerOptions = this.getCompilerOptions(projectDir, options).wait();
+			this.noEmitOnError = compilerOptions.noEmitOnError;
+			let typeScriptCompilerSettings = this.getTypeScriptCompiler().wait();
+			this.$logger.out(`Using tsc version ${typeScriptCompilerSettings.version}`.cyan);
+
+			// Core compilation
+			return this.runTranspilation(projectDir, typeScriptCompilerSettings.pathToCompiler, { compilerOptions }).wait();
+		}).future<string>()();
+	}
+
+	public getTypeScriptFiles(projectDir: string): IFuture<ITypeScriptFiles> {
+		return ((): ITypeScriptFiles => {
+			// Skip root's node_modules
+			let rootNodeModules = path.join(projectDir, "node_modules");
+			let projectFiles = this.$fs.enumerateFilesInDirectorySync(projectDir,
+				(fileName: string, fstat: IFsStats) => fileName !== rootNodeModules);
+			let typeScriptFiles = _.filter(projectFiles, file => path.extname(file) === ".ts");
+			let definitionFiles = _.filter(typeScriptFiles, file => _.endsWith(file, ".d.ts"));
+			return { definitionFiles: definitionFiles, typeScriptFiles: typeScriptFiles };
+		}).future<ITypeScriptFiles>()();
+	}
+
+	public isTypeScriptProject(projectDir: string): IFuture<boolean> {
+		return ((): boolean => {
+			let typeScriptFiles = this.getTypeScriptFiles(projectDir).wait();
+
+			if (typeScriptFiles.typeScriptFiles.length > typeScriptFiles.definitionFiles.length) { // We need this check because some of non-typescript templates(for example KendoUI.Strip) contain typescript definition files
+				return true;
+			}
+
+			return false;
+		}).future<boolean>()();
+	}
+
+	private hasTsConfigFile(projectDir: string): IFuture<boolean> {
+		return this.$fs.exists(this.getPathToTsConfigFile(projectDir));
+	}
+
+	private getPathToTsConfigFile(projectDir: string): string {
+		return path.join(projectDir, "tsconfig.json");
+	}
+
+	private getCompilerOptions(projectDir: string, options: ITypeScriptTranspileOptions): IFuture<ITypeScriptCompilerOptions> {
+		return ((): ITypeScriptCompilerOptions => {
+			let tsConfigFile: ITypeScriptConfig;
+			let pathToConfigJsonFile = this.getPathToTsConfigFile(projectDir);
+
+			if (this.hasTsConfigFile(projectDir).wait()) {
+				tsConfigFile = this.$fs.readJson(pathToConfigJsonFile).wait();
+			}
+
+			tsConfigFile = tsConfigFile || { compilerOptions: {} };
+			let compilerOptions = options.compilerOptions || {};
+			let defaultOptions = options.defaultCompilerOptions || {};
+
+			let compilerOptionsKeys = _.union(_.keys(compilerOptions), _.keys(tsConfigFile.compilerOptions), _.keys(defaultOptions));
+
+			let result: ITypeScriptCompilerOptions = {};
+			_.each(compilerOptionsKeys, (key: string) => {
+				// The order here is important.
+				result[key] = compilerOptions[key] || tsConfigFile.compilerOptions[key] || defaultOptions[key];
```

Comment:
The logic here is a bit off - if for example you have `compilerOptions = { noImplicitAny: false }` and `tsConfigFile = { noImplicitAny: true }`, the second one would override the first because you're getting the `compilerOptions[key]`'s value. Instead, you should be checking if `_.has(compilerOptions, key)` for example.